typescript-view-swapping - Daniel Yount

TheDanYount opened the pull request:
![Recording 2024-07-25 at 13 29 31](https://github.com/user-attachments/assets/1b6cad14-29ac-4c5f-8e5e-40d829f76c91)

TheDanYount committed:
Completed the exercise

brandon-moy commented on typescript-view-swapping/main.ts:
```diff
@@ -0,0 +1,25 @@
+
+function clickHandler(event: Event): void {
+  const eventTarget = event.target as HTMLElement;
+  if (eventTarget.matches('.tab')) {
+    for (const tab of document.querySelectorAll('.tab')) {
+      if (tab === eventTarget) {
```
We would want to query for these at the top of our `ts` file and store it in a variable. If we wait to query it inside of the event handler callback function, it would slow down our application because 1) we have to wait for the query here instead of when the page loads, and 2) we have to query _every_ time we click on a tab which means we're repeating work.

brandon-moy commented on typescript-view-swapping/main.ts:
```diff
@@ -0,0 +1,25 @@
+      } else {
+        tab.className = 'tab';
+      }
+    }
+    for (const view of document.querySelectorAll('.view')) {
+      if (!(view instanceof HTMLElement))
```
Same thing here, we would want to query for all of our `.view` elements at the top of the page.

brandon-moy commented on typescript-view-swapping/main.ts:
```diff
@@ -0,0 +1,25 @@
+$tabContainer.addEventListener('click', clickHandler);
+
+function clickHandler(event: Event): void {
+  const eventTarget = event.target as HTMLElement;
+  if (eventTarget.matches('.tab')) {
+    for (const tab of document.querySelectorAll('.tab')) {
```
I would recommend a guard clause here to easier readability and to exit the function early if the `eventTarget` is not what we want.

brandon-moy reviewed (changes requested):
Great job! There's a few places I think we can improve though.

TheDanYount committed:
Applied best practices

Did querySelectors ahead of time as to not slow down the page when buttons are clicked. Added guard statement. Added console log to guard statement so my %0^#^ing sytanx-checker will not consider it an empty code block).

TheDanYount committed:
Commented out console log

brandon-moy commented on typescript-view-swapping/main.ts:
```diff
@@ -0,0 +1,29 @@
+$tabContainer.addEventListener('click', clickHandler);
+
+function clickHandler(event: Event): void {
+  const eventTarget = event.target as HTMLElement;
+  if (!eventTarget.matches('.tab')) {
+    // console.log('Tab not clicked.');
+  } else {
+    for (const tab of $tabClassNodeList) {
```
This `if (!eventTarget.matches('.tab')` should `return` out of the function. Otherwise, if you have everything in an `else` block, it doesn't change how your code was previously.

The big thing to note about the guard clause is that it helps us by exiting a function early so that we do not need to look at any of the function code if our clause does not pass.

TheDanYount replied:
Whoops! Prettier didn't like it. I think I must have replaced it with that comment rather than simply adding that comment in before the return.

brandon-moy reviewed (changes requested):
The guard clause is missing a return!

TheDanYount committed:
Made sure guard statement is a guard statement

TheDanYount committed:
Removed else couching for post-guard-clause

brandon-moy reviewed (approved):
That's what I'm looking for, great job! 👍
